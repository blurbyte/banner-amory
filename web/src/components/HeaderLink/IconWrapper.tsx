import styled from 'styled-components';

const IconWrapper = styled.span`
  position: absolute;
  z-index: 9999;
  top: 0.8rem;
  left: -1.8rem;
`;

export default IconWrapper;
