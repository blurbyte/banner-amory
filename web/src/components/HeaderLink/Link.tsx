import { Link as BaseLink } from '@reach/router';
import styled from 'styled-components';

const Link = styled(BaseLink)`
  display: flex;
  width: 100%;
  height: 100%;
  padding: 0.4rem 1rem;
  align-items: flex-end;
  justify-content: flex-end;
  color: inherit;
  text-decoration: inherit;
  outline: none;
  border-top: 4px solid transparent;
  transition: border-top 0.15s linear 0.15s;
`;

export default Link;
