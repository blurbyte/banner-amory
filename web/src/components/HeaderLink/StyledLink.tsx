import { Link as BaseLink } from '@reach/router';
import styled from 'styled-components';

import { Colors } from '../../styles/colors';

const StyledLink = styled(BaseLink)`
  display: flex;
  position: relative;
  z-index: 9999;
  width: 12rem;
  height: 5rem;
  background: ${Colors.red};
  color: ${Colors.black};
  font-size: 1.8rem;
  font-weight: 700;
  text-decoration: none;
  margin-right: 3rem;
  padding: 0.4rem 1rem;
  align-items: flex-end;
  justify-content: flex-end;
  user-select: none;
`;

export default StyledLink;
