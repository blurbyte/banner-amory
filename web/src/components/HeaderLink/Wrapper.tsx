// wrapper is used here to avoid unknown props React warning

import styled from 'styled-components';
import { transparentize } from 'polished';

import { Color } from '@styles/color';
import Link from './Link';

type WrapperProps = {
  isActive?: boolean;
};

const Wrapper = styled.div<WrapperProps>`
  display: block;
  position: relative;
  z-index: 9999;
  width: 12rem;
  height: 5rem;
  background: ${props => (props.isActive ? Color.Red : transparentize(0.9, Color.Blue))};
  color: ${props => (props.isActive ? Color.Black : Color.Blue)};
  font-size: 1.8rem;
  font-weight: 700;
  text-decoration: none;
  margin-right: 3rem;
  user-select: none;
  transition: all 0.15s linear;

  & > ${Link}:focus {
    border-top: 4px solid ${props => (props.isActive ? Color.Black : Color.Blue)};
  }
`;

export default Wrapper;
