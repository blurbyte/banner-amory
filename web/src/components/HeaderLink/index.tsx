// header navigation link

import * as React from 'react';

import { Match } from '@reach/router';
import Link from './Link';
import IconWrapper from './IconWrapper';
import Wrapper from './Wrapper';

type LinkProps = {
  children: React.ReactNode;
  to: string;
  icon?: React.ComponentType<any>;
};

class HeaderLink extends React.Component<LinkProps> {
  render() {
    // using component injection pattern
    const { children, to, icon: InjectedIcon } = this.props;

    return (
      <Match path={to}>
        {({ match }) => (
          <Wrapper isActive={!!match} role="link">
            {InjectedIcon && (
              <IconWrapper>
                <InjectedIcon alternative={!match} />
              </IconWrapper>
            )}
            <Link to={to}>{children}</Link>
          </Wrapper>
        )}
      </Match>
    );
  }
}

export default HeaderLink;
