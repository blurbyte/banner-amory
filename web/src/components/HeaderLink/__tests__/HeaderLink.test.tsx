import * as React from 'react';
import { render, cleanup } from 'react-testing-library';

import HeaderLink from '../index';

beforeEach(cleanup);

test('renders correctly', () => {
  const Icon = () => <>Icon</>;
  const {container} = render(<HeaderLink icon={Icon} to="/test">Test</HeaderLink>);
  expect(container.firstChild).toMatchSnapshot();
});
