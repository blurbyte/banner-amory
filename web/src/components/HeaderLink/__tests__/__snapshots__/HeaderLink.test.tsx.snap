// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="sc-htpNat gnHPxh"
  role="link"
>
  <span
    class="sc-bwzfXH fvVgcj"
  >
    Icon
  </span>
  <a
    class="sc-bdVaJa Mjpsn"
    href="/test"
  >
    Test
  </a>
</div>
`;
